### PR TITLE
Use MaxLoopDevices variable from config file

### PR DESCRIPTION
**Pull request by locula**
## Description of the Pull Request (PR):

This PR makes use of the MaxLoopDevices variable defined in the config file in all the places where the maximum number of loop devices is referenced. This makes it easy for users to change this value as they choose.

### This fixes or addresses the following GitHub issues:

 - Addresses #5615

### CODE REVIEW

File: pkg/util/loop/loop_linux_test.go [hunk 1]
```diff
@@ -11,9 +11,27 @@ import (
 	"syscall"
 	"testing"
 
+	"github.com/sylabs/singularity/internal/pkg/buildcfg"
 	"github.com/sylabs/singularity/internal/pkg/test"
+	"github.com/sylabs/singularity/pkg/util/singularityconf"
 )
 
+func getMaxLoopDevices() int {
```
Comment (dtrudg) [inline pkg/util/loop/loop_linux_test.go]: This `getMaxLoopDevices` is repeated in many locations - please instead define it once as an exported/Public function (probably in `pkg/util/loop` is a good place - and then call it from other places it is needed.

File: pkg/util/crypt/crypt_dev.go [hunk 2]
```diff
@@ -36,11 +38,27 @@ var (
 	ErrInvalidPassphrase = errors.New("no key available with this passphrase")
 )
 
+func getMaxLoopDevices() int {
```
Comment (dtrudg) [inline pkg/util/crypt/crypt_dev.go]: This `getMaxLoopDevices` is repeated in many locations - please instead define it once as an exported/Public function (probably in `pkg/util/loop` is a good place - and then call it from other places it is needed.

File: pkg/ocibundle/tools/loop.go [hunk 3]
```diff
@@ -9,14 +9,32 @@ import (
 	"fmt"
 	"os"
 
+	"github.com/sylabs/singularity/internal/pkg/buildcfg"
 	"github.com/sylabs/singularity/pkg/util/loop"
+	"github.com/sylabs/singularity/pkg/util/singularityconf"
 )
 
+func getMaxLoopDevices() int {
```
Comment (dtrudg) [inline pkg/ocibundle/tools/loop.go]: This `getMaxLoopDevices` is repeated in many locations - please instead define it once as an exported/Public function (probably in `pkg/util/loop` is a good place - and then call it from other places it is needed.

File: internal/pkg/build/sources/packer_ext3_linux.go [hunk 4]
```diff
@@ -31,6 +33,22 @@ func (p *Ext3Packer) Pack(context.Context) (*types.Bundle, error) {
 	return p.b, nil
 }
 
+func getMaxLoopDevices() int {
```
Comment (dtrudg) [inline internal/pkg/build/sources/packer_ext3_linux.go]: This `getMaxLoopDevices` is repeated in many locations - please instead define it once as an exported/Public function (probably in `pkg/util/loop` is a good place - and then call it from other places it is needed.

**Review by dtrudg — CHANGES_REQUESTED** [review]

**Comment by dtrudg** [comment]
@locula - Apologies - in my comment before I didn't think through that these functions are bringing an `internal/` dependency into the supposedly reusable (from other projects) `pkg` code.

@cclerget - do you have any input on this? The general intent seems fine to me. However it does introduce the issue that things in `pkg/xxx` no depend on the `internal/pkg` config handling - so they aren't really independent of a Singularity install.

I think we probably need to look at how stuff could be shuffled around to avoid this hard `internal/` dep from `pkg/` if at all possible? I know we have #2316 which has been open for a long time to address this, and would hope we can make progress there rather than steps back.

**Comment by cclerget** [comment]
@dctrud @locula In order to deal with `buildcfg` dependency from a public package, the easiest approach is to create two files in `pkg/loop` and use go build tags :

- loop_singularity.go

```
// +build singularity_engine

package loop

import (
	"github.com/sylabs/singularity/internal/pkg/buildcfg"
	"github.com/sylabs/singularity/pkg/util/singularityconf"
)

func GetMaxLoopDevices() int {
	// if the caller has set the current config use it
	// otherwise parse the default configuration file
	cfg := singularityconf.GetCurrentConfig()
	if cfg == nil {
		var err error

		configFile := buildcfg.SINGULARITY_CONF_FILE
		cfg, err = singularityconf.Parse(configFile)
		if err != nil {
			return 256
		}
	}
	return int(cfg.MaxLoopDevices)
}
```

- loop_no_singularity.go

```
// +build !singularity_engine

package loop

func GetMaxLoopDevices() int {
        // externally imported package, use the default value
	return 256
}
```

This way external project would use `loop_no_singularity.go`.

> I think we probably need to look at how stuff could be shuffled around to avoid this hard internal/ dep from pkg/ if at all possible? I know we have #2316 which has been open for a long time to address this, and would hope we can make progress there rather than steps back.

Currently two packages are using `buildcfg` as indirect dependency: `pkg/build/types` and `pkg/util/crypt` and would require some modification before adding this check

**Review by cclerget — CHANGES_REQUESTED** [review]
Require to manage `buildcfg` internal dependency as described in https://github.com/hpcng/singularity/pull/5660#issuecomment-724054468

**Comment by locula** [comment]
@dctrud I've removed the internal dependencies and updated the code. Is there any additional feedback for this?

**Comment by dtrudg** [comment]
@locula - we'll come back to this soon after the 3.7.0 release. I'm afraid things are a bit busy in that regard at present.

**Comment by locula** [comment]
Hi, I just wanted to follow up and check if there are any updates on this.

**Review by cclerget — APPROVED** [review]
LGTM, thanks @locula ! And sorry for the delay

**Review by dtrudg — APPROVED** [review]

**Pull request merged**